readme.md
[![Build Status](https://travis-ci.org/CPFL/ouster.svg?branch=autoware_branch)](https://travis-ci.org/CPFL/ouster)

# OS1 ROS Node

## Building the ROS Node
* Supports Ubuntu 16.04 with ROS Kinetic (for ouster_driver)
* ROS installation instructions can be found [here](http://wiki.ros.org/kinetic/Installation/Ubuntu)
* Additionally requires ros-kinetic-pcl-ros and, optionally, ros-kinetic-rviz for visualization
* Be sure to source the ROS setup script before building. For example:`source /opt/ros/kinetic/setup.bash`
* Both packages can be built by catkin by moving them into a catkin workspace
* Build with `mkdir myworkspace && cd myworkspace && ln -s /path/to/ouster_driver ./src && catkin_make`

## Running the ROS Node
* Set up the ROS environment with `source /path/to/myworkspace/devel/setup.bash` in a new terminal for each command below
* For use with a running sensor:
  - To publish OS1 data as ROS topic `roslaunch ouster_driver os1.launch os1_hostname:=<os1_hostname> os1_udp_dest:=<udp_data_dest>` where `<os1_hostname>` can be the hostname or IP of the OS1 device and `<udp_data_dest>` is the IP to which the sensor should send data to (i.e., your computer's IP address on the interface connected to the OS1)
  - To record raw sensor output, run `rosbag record /os1_node/imu_packets /os1_node/lidar_packets` in another terminal
  - *Note: `os1_node/lidar_packets` and `os1_node/imu_packets` are the "raw data" topics, while `os1_node/points` is the ROS compatible XYZ topic and `os1_node/imu` is the ROS compatible IMU topic*
  - To visualize output, run `rviz -d /path/to/ouster_driver/rviz/viz.rviz` in another terminal
* For use with recorded sensor data:
  - To replay raw sensor output, run `roslaunch ouster_driver os1.launch replay:=true`
  - In a second terminal, run `rosbag play --clock <bagfile>`
  - To visualize output, run `rviz -d /path/to/ouster_driver/rviz/viz.rviz` in another terminal
* Sample raw sensor output is available [here](https://data.ouster.io/sample-data-2018-08-29)

## OS1 address configuration
* By default, the OS1 uses DHCP to obtain its IP address (the `<os1_hostname>` parameter above). The **"OS-1-64/16 High Resolution Imaging LIDAR: Software User Guide"** by  Ouster Inc. recommends using `dnsmasq`. 
* Setting IP address dynamically (default)
  - Install `dnsmasq` (when required), as `sudo apt install dnsmasq dnsmasq-utils`
  - Identify the interface name the OS1 is attached to. If this interface is not configured yet, setup its IP address and network mask. Use the NetworkManager and use "manual" in IPv4 settings, or use the `ifconfig` command (ex., `ifconfig <INTERFACENAME> 192.168.2.1 netmask 255.255.255.0 up`)
  - Modify the configuration file `/etc/dnsmasq.conf`, define/modify the `interface=` line with the interface name above. Also define the `dhcp-range=` value according to your interface (ex., `dhcp-range=192.168.2.50,192.168.2.150,12h`, in accordance to the `ifconfig` example above)
  - Start (stop) `dnsmasq` system service (`sudo systemctl stop dnsmasq` to stop, `sudo systemctl start dnsmasq` to start)
  - Identify the address assigned to the OS1 using `journalctl -fu dnsmasq`. Example:
```
   Oct 23 09:45:56 <HOSTNAME> dnsmasq-dhcp[30010]: DHCPREQUEST(<INTERFACENAME>) 192.168.2.*** bc:0f:**:**:**:**
   Oct 23 09:45:56 <HOSTNAME> dnsmasq-dhcp[30010]: DHCPACK(<INTERFACENAME>) 192.168.2.*** bc:0f:**:**:**:** os1-XXXXXXXXXXXX
```
where `<HOSTNAME>` is your computer's assigned hostname, `<INTERFACENAME>` is the interface configured above, `192.168.2.***` is the IP address assigned to the sensor, `bc:0f:**:**:**:**` is the OS1 MAC address and `os1-XXXXXXXXXXXX` is the sensor's hostname (`XXXXXXXXXXXX` is replaced by the sensor's serial number). *Note: instead of "\*" an actual number will be shown.*

  - You can set `<os1_hostname>` as either the assigned IP address (ex., `192.168.2.***`) or the sensor's hostname (ex., `os1-XXXXXXXXXXXX`).

## OS1 mode configuration
* Setting the LiDAR horizontal resolution and scan rate
  - Please identify the current sensor IP address or hostname using the steps above
  - The supported values for `lidar_mode` are: 512x10, 1024x10, 2048x10, 512x20, 1024x20. The first value is horizontal resolution and the second is scan rate.
  - Connect to the sensor and reconfigure the `lidar_mode` parameter, storing the modification internally so to use it at next boot. You can skip the `write_config_txt` command to use default value at next boot. `os1-XXXXXXXXXXXX` is the sensor's hostname (`XXXXXXXXXXXX` is replaced by the sensor's serial number).
```bash
  nc os1-XXXXXXXXXXXX 7501
  get_config_param active lidar_mode
  set_config_param lidar_mode 2048x10
  write_config_txt
  reinitialize
```

##Todo
- [x] Velodyne compatibility mode.
- [x] Configure active sensor parameters from driver code, in particular `lidar_mode`.
- [ ] Fix `replay` option currently not working.
- [ ] Verify the sensor operation information and attempt software reset on error condition, and current parameter values to avoid unnecessary rewrites/reinitialization.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package ouster
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------
* Sensor model and firmware version validation
* Added sensor model and firmware version validation. The driver currently only works for OS-1-64 and on firmware version 1.10 some configuration parameters disappeared.
* Parameter validation and new pointcloud types
* Added parameter validation to avoid reinitializing the LiDAR. Added new pointcloud types XYZIRF and XYZIRFN. Changed pulse_mode parameter from integer to string for readability. Updated the readme file.
* Contributors: Alexander Carballo, alexandrx

0.1.6 (2019-03-13)
------------------
* Renamed ouster_ros to ouster_driver

0.1.5 (2019-03-13)
------------------
* updated package.xml

0.1.4 (2019-03-13)
------------------
* Updated project name to follow ros standards

0.1.3 (2019-03-06)
------------------
* Updated LICENSE

0.1.2 (2019-03-06)
------------------
* Changed operation mode from integer to string to ease reading.
* New pointcloud mode format added
* Improved parameter documentation
* removed the old bool pointcloud_mode_xyzir parameter (to match Velodyne's mode) and replaced with a way to specify the desired pointcloud format, several options added including Ouster's native pointcloud format.
* Added install commands
* run CI on kinetic and melodic only
* Fixed launch file to deal with hidden parameters, fixed readme.md to update the TODO, fixed os1_node.cpp to reduce by half the scan duration variable in case of 20Hz operation.
* Updated the driver to support advanced parameter configuration (lidar-mode, pulse-mode and window-rejection-enable)
* Eliminated the ouster-ros static library 
* Updated readme.md incorrect info for lidar address configuration.
* Switched to industrial-ci
* First commit for Autoware version branch. Modified the original Ouster ROS code to suit Autoware needs. Code structure is more simple, only ROS driver code is kept, and the readme.md file has more details on how to configure the sensor and use this driver. To further ease Autoware integration, Velodyne compatibility options were added and tested with nd-_matching, ray-ground-filter and euclidean-cluster nodes.
* Improve positional argument parsing in viz
* Use allocate-shared for Eigen compatibility
* Support building visualizer with newer VTK
* Readme tweaks and new example data
* Update readme.md
* updated the readme to include viz in the contents
* Initial commit of the example visualizer
* Uses VTK6 to display point clouds and range/intensity/noise images
* No dependencies other than VTK6, Eigen3, and a C++11 compiler
* Currently only supports linux; tested on multiple distributions and platforms
* Use correct units in sensor-msgs::Imu
* Deal with time going backwards in replays
* Contributors: Alexander Carballo, Dima Garbuzov, Roy Rinberg, alexandrx, amc-nu, hooic, rafficm, rrinberg


src/os1_ros.cpp
#include <pcl_conversions/pcl_conversions.h>

#include "ouster/os1.h"
#include "ouster/os1_packet.h"
#include "ouster_driver/os1_ros.h"

namespace ouster_driver {
namespace OS1 {

using namespace ouster::OS1;

static std::string _pointcloud_mode = "NATIVE";

ns timestamp_of_imu_packet(const PacketMsg& pm) {
    return ns(imu_gyro_ts(pm.buf.data()));
}

ns timestamp_of_lidar_packet(const PacketMsg& pm) {
    return ns(col_timestamp(nth_col(0, pm.buf.data())));
}

bool read_imu_packet(const client& cli, PacketMsg& m) {
    m.buf.resize(imu_packet_bytes + 1);
    return read_imu_packet(cli, m.buf.data());
}

bool read_lidar_packet(const client& cli, PacketMsg& m) {
    m.buf.resize(lidar_packet_bytes + 1);
    return read_lidar_packet(cli, m.buf.data());
}

sensor_msgs::Imu packet_to_imu_msg(const PacketMsg& p, const std::string& frame) {
    const double standard_g = 9.80665;
    sensor_msgs::Imu m;
    const uint8_t* buf = p.buf.data();

    m.header.stamp.fromNSec(imu_gyro_ts(buf));
    /**
     * @note Modified to support custom message frame name
     */
    //m.header.frame_id = "os1_imu"; // original code of OS1 driver
    m.header.frame_id = frame; //using defined message frame name

    m.orientation.x = 0;
    m.orientation.y = 0;
    m.orientation.z = 0;
    m.orientation.w = 0;

    m.linear_acceleration.x = imu_la_x(buf) * standard_g;
    m.linear_acceleration.y = imu_la_y(buf) * standard_g;
    m.linear_acceleration.z = imu_la_z(buf) * standard_g;

    m.angular_velocity.x = imu_av_x(buf) * M_PI / 180.0;
    m.angular_velocity.y = imu_av_y(buf) * M_PI / 180.0;
    m.angular_velocity.z = imu_av_z(buf) * M_PI / 180.0;

    for (int i = 0; i < 9; i++) {
        m.orientation_covariance[i] = -1;
        m.angular_velocity_covariance[i] = 0;
        m.linear_acceleration_covariance[i] = 0;
    }
    for (int i = 0; i < 9; i += 4) {
        m.linear_acceleration_covariance[i] = 0.01;
        m.angular_velocity_covariance[i] = 6e-4;
    }

    return m;
}

sensor_msgs::PointCloud2 cloud_to_cloud_msg(const CloudOS1& cloud, ns timestamp,
                                            const std::string& frame) {
    sensor_msgs::PointCloud2 msg;
    
    //pcl::toROSMsg(cloud, msg); //<-- original code of OS1 driver
    /**
     * @note Added to support Velodyne compatible pointcloud format for Autoware
     */
    if (_pointcloud_mode == "XYZ") {
    	CloudOS1XYZ cloud_aux;
    	convert2XYZ(cloud, cloud_aux);
    	pcl::toROSMsg(cloud_aux, msg);
    } else if (_pointcloud_mode == "XYZI") {
    	CloudOS1XYZI cloud_aux;
    	convert2XYZI(cloud, cloud_aux);
    	pcl::toROSMsg(cloud_aux, msg);
    } else if (_pointcloud_mode == "XYZIR") {
    	CloudOS1XYZIR cloud_aux;
    	convert2XYZIR(cloud, cloud_aux);
    	pcl::toROSMsg(cloud_aux, msg);
    } else if (_pointcloud_mode == "XYZIF") {
    	CloudOS1XYZIF cloud_aux;
    	convert2XYZIF(cloud, cloud_aux);
    	pcl::toROSMsg(cloud_aux, msg);
    } else if (_pointcloud_mode == "XYZIFN") {
    	CloudOS1XYZIFN cloud_aux;
    	convert2XYZIFN(cloud, cloud_aux);
    	pcl::toROSMsg(cloud_aux, msg);
    } else { //"NATIVE"
    	pcl::toROSMsg(cloud, msg);
    }

    msg.header.frame_id = frame;
    /**
     * @note Changed timestamp from LiDAR to ROS time for Autoware operation
     */
    //msg.header.stamp.fromNSec(timestamp.count()); //<-- original code of OS1 driver
    msg.header.stamp = ros::Time::now();  //<-- prefered time mode in Autoware
    return msg;
}

static PointOS1 nth_point(int ind, const uint8_t* col_buf) {
    float h_angle_0 = col_h_angle(col_buf);
    auto tte = trig_table[ind];
    const uint8_t* px_buf = nth_px(ind, col_buf);
    float r = px_range(px_buf) / 1000.0;
    float h_angle = tte.beam_azimuth_angles + h_angle_0;

    PointOS1 point;
    point.reflectivity = px_reflectivity(px_buf);
    point.intensity = px_signal_photons(px_buf);
    point.noise = px_noise_photons(px_buf); //added to extract ambient noise data
    point.x = -r * tte.cos_beam_altitude_angles * cosf(h_angle);
    point.y = r * tte.cos_beam_altitude_angles * sinf(h_angle);
    point.z = r * tte.sin_beam_altitude_angles;
    point.ring = ind;

    return point;
}
void add_packet_to_cloud(ns scan_start_ts, ns scan_duration,
                         const PacketMsg& pm, CloudOS1& cloud) {
    const uint8_t* buf = pm.buf.data();
    for (int icol = 0; icol < columns_per_buffer; icol++) {
        const uint8_t* col_buf = nth_col(icol, buf);
        float ts = (col_timestamp(col_buf) - scan_start_ts.count()) /
                   (float)scan_duration.count();

        for (int ipx = 0; ipx < pixels_per_column; ipx++) {
            auto p = nth_point(ipx, col_buf);
            p.t = ts;
            cloud.push_back(p);
        }
    }
}

void spin(const client& cli,
          const std::function<void(const PacketMsg& pm)>& lidar_handler,
          const std::function<void(const PacketMsg& pm)>& imu_handler) {
    PacketMsg lidar_packet, imu_packet;
    lidar_packet.buf.resize(lidar_packet_bytes + 1);
    imu_packet.buf.resize(imu_packet_bytes + 1);

    while (ros::ok()) {
        auto state = poll_client(cli);
        if (state & ERROR) {
            ROS_ERROR("spin: poll_client returned error");
            return;
        }
        if (state & LIDAR_DATA) {
            if (read_lidar_packet(cli, lidar_packet.buf.data()))
                lidar_handler(lidar_packet);
        }
        if (state & IMU_DATA) {
            if (read_imu_packet(cli, imu_packet.buf.data()))
                imu_handler(imu_packet);
        }
        ros::spinOnce();
    }
}

static ns nearest_scan_dur(ns scan_dur, ns ts) {
    return ns((ts.count() / scan_dur.count()) * scan_dur.count());
};

std::function<void(const PacketMsg&)> batch_packets(
    ns scan_dur, const std::function<void(ns, const CloudOS1&)>& f) {
    auto cloud = std::make_shared<OS1::CloudOS1>();
    auto scan_ts = ns(-1L);

    return [=](const PacketMsg& pm) mutable {
        ns packet_ts = OS1::timestamp_of_lidar_packet(pm);
        if (scan_ts.count() == -1L)
            scan_ts = nearest_scan_dur(scan_dur, packet_ts);

        OS1::add_packet_to_cloud(scan_ts, scan_dur, pm, *cloud);

        auto batch_dur = packet_ts - scan_ts;
        if (batch_dur >= scan_dur || batch_dur < ns(0)) {
            f(scan_ts, *cloud);

            cloud->clear();
            scan_ts = ns(-1L);
        }
    };
}

/**
 * @note Added to support Velodyne compatible pointcloud format for Autoware
 */
void set_point_mode(std::string mode_xyzir)
{
    _pointcloud_mode = mode_xyzir;
}

void convert2XYZ(const CloudOS1& in, CloudOS1XYZ& out) 
{
   out.points.clear();
   pcl::PointXYZ q;
   for (auto p : in.points) {
       q.x = p.x;
       q.y = p.y;
       q.z = p.z;
       out.points.push_back(q);
   }
}

void convert2XYZI(const CloudOS1& in, CloudOS1XYZI& out) 
{
   out.points.clear();
   pcl::PointXYZI q;
   for (auto p : in.points) {
       q.x = p.x;
       q.y = p.y;
       q.z = p.z;
       q.intensity = p.intensity;
       out.points.push_back(q);
   }
}

/**
 * @note Added to support Velodyne compatible pointcloud format for Autoware
 */
void convert2XYZIR(const CloudOS1& in, CloudOS1XYZIR& out) 
{
   out.points.clear();
   PointXYZIR q;
   for (auto p : in.points) {
       q.x = p.x;
       q.y = p.y;
       q.z = p.z;
       q.intensity = ((float)p.intensity/65535.0)*255.0; //velodyne uses values in [0..255] range
       q.ring = p.ring;
       out.points.push_back(q);
   }
}

/**
 * @note Extract intensity and reflectivity data
 */
void convert2XYZIF(const CloudOS1& in, CloudOS1XYZIF& out) 
{
   out.points.clear();
   PointXYZIF q;
   for (auto p : in.points) {
       q.x = p.x;
       q.y = p.y;
       q.z = p.z;
       q.intensity = p.intensity;
       q.reflectivity = p.reflectivity;
       out.points.push_back(q);
   }
}

/**
 * @note Extract intensity and reflectivity and ambient noise data
 */

void convert2XYZIFN(const CloudOS1& in, CloudOS1XYZIFN& out) 
{
   out.points.clear();
   PointXYZIFN q;
   for (auto p : in.points) {
       q.x = p.x;
       q.y = p.y;
       q.z = p.z;
       q.intensity = p.intensity;
       q.reflectivity = p.reflectivity;
       q.noise = p.noise;
       out.points.push_back(q);
   }
}

}
}


src/os1_node.cpp
/**
 * Example node to publish OS1 output on ROS topics
 *
 * Additionally, this node can be used to record and replay raw sesnor output
 * by publishing and listening to PacketMsg topics
 *
 * ROS Parameters
 * scan_dur_ns: nanoseconds to batch lidar packets before publishing a cloud
 * os1_hostname: hostname or IP in dotted decimal form of the sensor
 * os1_udp_dest: hostname or IP where the sensor will send data packets
 * os1_lidar_port: port to which the sensor should send lidar data
 * os1_imu_port: port to which the sensor should send imu data
 * replay_mode: when true, the node will listen on ~/lidar_packets and
 *   ~/imu_packets for data instead of attempting to connect to a sensor
 */

#include <chrono>
#include <functional>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

#include <pcl_conversions/pcl_conversions.h>
#include <ros/console.h>
#include <ros/ros.h>
#include <sensor_msgs/Imu.h>
#include <sensor_msgs/PointCloud2.h>

#include "ouster_driver/PacketMsg.h"
#include "ouster_driver/os1_ros.h"

using ns = std::chrono::nanoseconds;
using PacketMsg = ouster_driver::PacketMsg;

int main(int argc, char** argv) {
    ros::init(argc, argv, "os1_node");
    ros::NodeHandle nh("~");

    auto scan_dur = ns(nh.param("scan_dur_ns", 100000000));
    auto os1_hostname = nh.param("os1_lidar_address", std::string("localhost"));
    auto os1_udp_dest = nh.param("pc_address", std::string("192.168.1.1"));
    auto os1_lidar_port = nh.param("os1_lidar_port", -1);
    auto os1_imu_port = nh.param("os1_imu_port", -1);
    auto replay_mode = nh.param("replay", true);
    auto points_topic_name = nh.param("points_topic_name", std::string("/points_raw"));
    auto imu_topic_name = nh.param("imu_topic_name", std::string("/imu_raw"));
    auto lidar_frame_name = nh.param("lidar_frame_name", std::string("/os1"));
    auto imu_frame_name = nh.param("imu_frame_name", std::string("/os1_imu"));
    auto pointcloud_mode = nh.param("pointcloud_mode", std::string("NATIVE"));
    auto operation_mode_str = nh.param("operation_mode", std::string("1024x10"));
    auto pulse_mode = nh.param("pulse_mode", 0);
    auto window_rejection = nh.param("window_rejection", true);
    
    //0:512x10, 1:1024x10, 2:2048x10, 3:512x20, 4:1024x20
    ouster::OS1::OperationMode operation_mode = ouster::OS1::MODE_1024x10; // default mode
    if (operation_mode_str == std::string("512x10")) {
    	operation_mode = ouster::OS1::MODE_512x10;
    } else if (operation_mode_str == std::string("1024x10")) {
    	operation_mode = ouster::OS1::MODE_1024x10;
    } else if (operation_mode_str == std::string("2048x10")) {
    	operation_mode = ouster::OS1::MODE_2048x10;
    } else if (operation_mode_str == std::string("512x20")) {
    	operation_mode = ouster::OS1::MODE_512x20;
    } else if (operation_mode_str == std::string("1024x20")) {
    	operation_mode = ouster::OS1::MODE_1024x20;
    } else {
    	ROS_WARN_STREAM("Operation mode: \"" << operation_mode_str << "\" is not supported, reverting to default mode \"1024x10\"");
    }
    
    /**
     * @note Added to support Velodyne compatible pointcloud format for Autoware
     */
    //defines the pointcloud mode
    ouster_driver::OS1::set_point_mode(pointcloud_mode);
    //----------------
    /**
     * @note Added to support advanced mode parameters configuration for Autoware
     */
    //defines the advanced parameters
    ouster::OS1::set_advanced_params(operation_mode, (ouster::OS1::PulseMode)pulse_mode, window_rejection);
    auto queue_size = 10;
    if (operation_mode == ouster::OS1::MODE_512x20 || operation_mode == ouster::OS1::MODE_1024x20) {
    	queue_size = 20;
    	scan_dur = scan_dur / 2; //scan duration should be smaller at faster frame rates
    }
    //----------------

    auto lidar_pub = nh.advertise<sensor_msgs::PointCloud2>(points_topic_name, queue_size);
    auto imu_pub = nh.advertise<sensor_msgs::Imu>(imu_topic_name, queue_size);

    auto lidar_handler = ouster_driver::OS1::batch_packets(
        scan_dur, [&](ns scan_ts, const ouster_driver::OS1::CloudOS1& cloud) {
            lidar_pub.publish(
              ouster_driver::OS1::cloud_to_cloud_msg(cloud, scan_ts, lidar_frame_name));
        });

    auto imu_handler = [&](const PacketMsg& p) {
        imu_pub.publish(ouster_driver::OS1::packet_to_imu_msg(p, imu_frame_name));
    };

    if (replay_mode) {
        auto lidar_packet_sub = nh.subscribe<PacketMsg, const PacketMsg&>(
            "lidar_packets", 500, lidar_handler);
        auto imu_packet_sub = nh.subscribe<PacketMsg, const PacketMsg&>(
            "imu_packets", 500, imu_handler);
        ros::spin();
    } else {
        auto lidar_packet_pub = nh.advertise<PacketMsg>("lidar_packets", 500);
        auto imu_packet_pub = nh.advertise<PacketMsg>("imu_packets", 500);

        auto cli = ouster::OS1::init_client(os1_hostname, os1_udp_dest,
                                            os1_lidar_port, os1_imu_port);
        if (!cli) {
            ROS_ERROR("Failed to initialize sensor at: %s", os1_hostname.c_str());
            return 1;
        }

        ouster_driver::OS1::spin(*cli,
                              [&](const PacketMsg& pm) {
                                  lidar_packet_pub.publish(pm);
                                  lidar_handler(pm);
                              },
                              [&](const PacketMsg& pm) {
                                  imu_packet_pub.publish(pm);
                                  imu_handler(pm);
                              });
    }
    return 0;
}


src/os1.cpp
#include <cerrno>
#include <chrono>
#include <cstdio>
#include <cstring>
#include <iostream>
#include <memory>

#include <arpa/inet.h>
#include <netdb.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>

#include "ouster/os1.h"

namespace ouster {
namespace OS1 {

using ns = std::chrono::nanoseconds;

/**
 * @note Added to support advanced mode parameters configuration for Autoware
 */
static OperationMode _operation_mode = ouster::OS1::MODE_1024x10;
static PulseMode _pulse_mode = ouster::OS1::PULSE_STANDARD;
static bool _window_rejection = true;
static std::string _operation_mode_str = "";
static std::string _pulse_mode_str = "";
static std::string _window_rejection_str = "";
//----------------

struct client {
    int lidar_fd;
    int imu_fd;
    ~client() {
        close(lidar_fd);
        close(imu_fd);
    }
};

static int udp_data_socket(int port) {
    int sock_fd = socket(PF_INET, SOCK_DGRAM, 0);
    if (sock_fd < 0) {
        std::cerr << "socket: " << std::strerror(errno) << std::endl;
        return -1;
    }

    struct sockaddr_in my_addr;
    memset((char*)&my_addr, 0, sizeof(my_addr));
    my_addr.sin_family = AF_INET;
    my_addr.sin_port = htons(port);
    my_addr.sin_addr.s_addr = INADDR_ANY;

    if (bind(sock_fd, (struct sockaddr*)&my_addr, sizeof(my_addr)) < 0) {
        std::cerr << "bind: " << std::strerror(errno) << std::endl;
        return -1;
    }

    struct timeval timeout;
    timeout.tv_sec = 1;
    timeout.tv_usec = 0;
    if (setsockopt(sock_fd, SOL_SOCKET, SO_RCVTIMEO, &timeout,
                   sizeof(timeout)) < 0) {
        std::cerr << "setsockopt: " << std::strerror(errno) << std::endl;
        return -1;
    }

    return sock_fd;
}

static int cfg_socket(const char* addr) {
    struct addrinfo hints, *info_start, *ai;

    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;

    int ret = getaddrinfo(addr, "7501", &hints, &info_start);
    if (ret != 0) {
        std::cerr << "getaddrinfo: " << gai_strerror(ret) << std::endl;
        return -1;
    }
    if (info_start == NULL) {
        std::cerr << "getaddrinfo: empty result" << std::endl;
        return -1;
    }

    int sock_fd;
    for (ai = info_start; ai != NULL; ai = ai->ai_next) {
        sock_fd = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
        if (sock_fd < 0) {
            std::cerr << "socket: " << std::strerror(errno) << std::endl;
            continue;
        }

        if (connect(sock_fd, ai->ai_addr, ai->ai_addrlen) == -1) {
            close(sock_fd);
            continue;
        }

        break;
    }

    freeaddrinfo(info_start);
    if (ai == NULL) {
        return -1;
    }

    return sock_fd;
}

std::shared_ptr<client> init_client(const std::string& hostname,
                                    const std::string& udp_dest_host,
                                    int lidar_port, int imu_port) {
    int sock_fd = cfg_socket(hostname.c_str());

    if (sock_fd < 0) return std::shared_ptr<client>();

    std::string cmd;

    auto do_cmd = [&](const std::string& op, const std::string& val) {
        const size_t max_res_len = 64;
        char read_buf[max_res_len + 1];

        ssize_t len;
        std::string cmd = op + " " + val + "\n";

        len = write(sock_fd, cmd.c_str(), cmd.length());
        if (len != (ssize_t)cmd.length()) {
            //std::cerr << "init_client: failed to send command" << std::endl;
            return std::string("[error] init_client: failed to send command");
        }

        len = read(sock_fd, read_buf, max_res_len);
        if (len < 0) {
            //std::cerr << "read: " << std::strerror(errno) << std::endl;
            return std::string("[error] read: ") + std::strerror(errno);
        }
        read_buf[len] = '\0';

        auto res = std::string(read_buf);
        res.erase(res.find_last_not_of(" \r\n\t") + 1);
        
        return res;
    };
    
    auto do_cmd_chk = [&](const std::string& op, const std::string& val) {
    	auto res = do_cmd(op, val);
    	if (res.find("[error]") != std::string::npos) {
    		std::cerr << res << std::endl;
    		return false;
    	} else if (res != op) {
            std::cerr << "init_client: command \"" << op << " " << val << "\" failed with \""
                      << res << "\"" << std::endl;
            return false;
        }
        return true;
    };
    
    auto get_cmd = [&](const std::string& op, const std::string& val) {
    	auto res = do_cmd(op, val);
    	if (res.find("error") != std::string::npos) {
    		std::cerr << res << std::endl;
    		return std::string("");
    	} 
        return res;
    };

    bool success = true;
    success &= do_cmd_chk("set_udp_port_lidar", std::to_string(lidar_port));
    success &= do_cmd_chk("set_udp_port_imu", std::to_string(imu_port));
    success &= do_cmd_chk("set_udp_ip", udp_dest_host);

    if (!success) return std::shared_ptr<client>();
    
    /**
     * @note Added to support advanced mode parameters configuration
     */
    //read the current settings
    auto curr_operation_mode_str = get_cmd("get_config_param", "active lidar_mode");
    auto curr_pulse_mode_str = get_cmd("get_config_param", "active pulse_mode");
    auto curr_window_rejection_str = get_cmd("get_config_param", "active window_rejection_enable");
    bool do_configure = false;
    success = true;
    
    //setting advanced mode parameters (if necessary)
    if (curr_operation_mode_str != _operation_mode_str) {
    	success &= do_cmd_chk("set_config_param", "lidar_mode " + _operation_mode_str);
    	do_configure = true;
   	}
   	if (curr_pulse_mode_str != _pulse_mode_str) {
	    success &= do_cmd_chk("set_config_param", "pulse_mode " + _pulse_mode_str);
	    do_configure = true;
    }
    if (curr_window_rejection_str != _window_rejection_str) {
		success &= do_cmd_chk("set_config_param", "window_rejection_enable " + _window_rejection_str);
		do_configure = true;
   	}
    
    if (!success) return std::shared_ptr<client>();
    
    //reinitialize to take effect (if necessary)
	if (do_configure) {
		success &= do_cmd_chk("reinitialize", "");
		if (!success) return std::shared_ptr<client>();
		std::cout << "Parameters configured, reinitializing sensor" << std::endl;
   	} else {
   		std::cout << "Parameters already configured, no need to reinitialize" << std::endl;
   	}
    //----------------
    
    close(sock_fd);

    int lidar_fd = udp_data_socket(lidar_port);
    int imu_fd = udp_data_socket(imu_port);
    auto cli = std::make_shared<client>();
    cli->lidar_fd = lidar_fd;
    cli->imu_fd = imu_fd;
    return cli;
}

client_state poll_client(const client& c) {
    fd_set rfds;
    FD_ZERO(&rfds);
    FD_SET(c.lidar_fd, &rfds);
    FD_SET(c.imu_fd, &rfds);

    int max_fd = std::max(c.lidar_fd, c.imu_fd);

    int retval = select(max_fd + 1, &rfds, NULL, NULL, NULL);

    client_state res = client_state(0);
    if (retval == -1) {
        std::cerr << "select: " << std::strerror(errno) << std::endl;
        res = client_state(res | ERROR);
    } else if (retval) {
        if (FD_ISSET(c.lidar_fd, &rfds)) res = client_state(res | LIDAR_DATA);
        if (FD_ISSET(c.imu_fd, &rfds)) res = client_state(res | IMU_DATA);
    }
    return res;
}

static bool recv_fixed(int fd, void* buf, size_t len) {
    ssize_t n = recvfrom(fd, buf, len + 1, 0, NULL, NULL);
    if (n == (ssize_t)len)
        return true;
    else if (n == -1)
        std::cerr << "recvfrom: " << std::strerror(errno) << std::endl;
    else
        std::cerr << "Unexpected udp packet length: " << n << std::endl;
    return false;
}

bool read_lidar_packet(const client& cli, uint8_t* buf) {
    return recv_fixed(cli.lidar_fd, buf, lidar_packet_bytes);
}

bool read_imu_packet(const client& cli, uint8_t* buf) {
    return recv_fixed(cli.imu_fd, buf, imu_packet_bytes);
}

/**
 * @note Added to support advanced mode parameters configuration for Autoware
 */
//----------------
void set_advanced_params(std::string operation_mode_str, std::string pulse_mode_str, bool window_rejection)
{
   _operation_mode_str = operation_mode_str;
   _pulse_mode_str = pulse_mode_str;
   _window_rejection = window_rejection;
   
   _operation_mode = ouster::OS1::MODE_1024x10;
   if (_operation_mode_str == std::string("512x10")) {
       _operation_mode = ouster::OS1::MODE_512x10;
   } else if (_operation_mode_str == std::string("1024x10")) {
       _operation_mode = ouster::OS1::MODE_1024x10;
   } else if (_operation_mode_str == std::string("2048x10")) {
       _operation_mode = ouster::OS1::MODE_2048x10;
   } else if (_operation_mode_str == std::string("512x20")) {
       _operation_mode = ouster::OS1::MODE_512x20;
   } else if (_operation_mode_str == std::string("1024x20")) {
       _operation_mode = ouster::OS1::MODE_1024x20;
   } else {
   	   std::cout << "Selected operation mode " << _operation_mode_str << " is invalid, using default mode \"1024x10\"" << std::endl;
   }

   _pulse_mode = ouster::OS1::PULSE_STANDARD;
   if (_pulse_mode_str == std::string("STANDARD")) {
   	   _pulse_mode = ouster::OS1::PULSE_STANDARD;
   } else if (_pulse_mode_str == std::string("NARROW")) {
   	   _pulse_mode = ouster::OS1::PULSE_NARROW;
   } else {
   	   std::cout << "Selected pulse mode " << _pulse_mode_str << " is invalid, using default mode \"STANDARD\"" << std::endl;
   }

   _window_rejection_str = ((_window_rejection) ? std::string("1") : std::string("0"));
}
//----------------

}
}
